### PR TITLE
Add governance Info action test

### PR DESCRIPTION
**Pull request by carbolymer**
# Description

Add a governance info testnet scenario.

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated.  These may include:
  - golden tests
  - property tests
  - roundtrip tests
  - integration tests
  See [Runnings tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [ ] Any changes are noted in the `CHANGELOG.md` for affected package
- [ ] The version bounds in `.cabal` files are updated
- [ ] CI passes. See note on CI.  The following CI checks are required:
  - [ ] Code is linted with `hlint`.  See `.github/workflows/check-hlint.yml` to get the `hlint` version
  - [ ] Code is formatted with `stylish-haskell`.  See `.github/workflows/stylish-haskell.yml` to get the `stylish-haskell` version
  - [ ] Code builds on Linux, MacOS and Windows for `ghc-8.10.7` and `ghc-9.2.7`
- [ ] Self-reviewed the diff

# Note on CI
If your PR is from a fork, the necessary CI jobs won't trigger automatically for security reasons.
You will need to get someone with write privileges.  Please contact IOG node developers to do this
for you.

### CODE REVIEW

File: cardano-testnet/test/cardano-testnet-test/Cardano/Testnet/Test/LedgerEvents/Gov/InfoAction.hs [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+              map snd
+              $ filter ((== govActionIdx) . fst)
+              $ map (first (\gai -> do
+                  let (L.GovActionIx i) = L.gaidGovActionIx gai
+                  i
+                ))
+              $ Map.assocs $ L.proposalsGovActionStates proposals
+        case actions of
+          [L.GovActionState{L.gasDRepVotes=votes}]
+            | length votes == 3 -> ias{hasReceivedVotes = True}
```
Comment (smelc) [inline cardano-testnet/test/cardano-testnet-test/Cardano/Testnet/Test/LedgerEvents/Gov/InfoAction.hs:404]: Quite readable for what it does, nice :heart:

File: cardano-testnet/test/cardano-testnet-test/Cardano/Testnet/Test/LedgerEvents/Gov/InfoAction.hs [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+filterNewGovProposals _ _ = False
+
+-- | Fold accumulator for checking action state
+data InfoActionState = InfoActionState
+  { hasReceivedVotes :: !Bool
+  , isRemoved :: !Bool
+  } deriving Show
+
+foldBlocksCheckInfoAction
+  :: FilePath
```
Comment (smelc) [inline cardano-testnet/test/cardano-testnet-test/Cardano/Testnet/Test/LedgerEvents/Gov/InfoAction.hs:369]:
```suggestion
  :: FilePath -- ^ Where to store debug logs
```

**Review by smelc — APPROVED** [review]
Compared with the existing `cardano-testnet/test/cardano-testnet-test/Cardano/Testnet/Test/LedgerEvents/Gov/ProposeNewConstitution.hs` and this one works similarly :+1:

One question though: don't we want to use `query gov-state` (or alike) instead of using `foldBlocks` at the end? I thought we were moving away from this `foldBlock` pattern, in its existing shape at least.

File: cardano-testnet/test/cardano-testnet-test/Cardano/Testnet/Test/LedgerEvents/Gov/InfoAction.hs [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+    , "--address", Text.unpack $ paymentKeyInfoAddr $ head wallets
+    , "--cardano-mode"
+    , "--testnet-magic", show @Int testnetMagic
+    , "--out-file", work </> "utxo-1.json"
+    ]
+
+  txin1 <- do
+    utxoJson <- H.leftFailM . H.readJsonFile $ work </> "utxo-1.json"
+    UTxO utxo <- H.noteShowM $ decodeEraUTxO sbe utxoJson
+    H.noteShow =<< H.headM (Map.keys utxo)
```
Comment (smelc) [inline cardano-testnet/test/cardano-testnet-test/Cardano/Testnet/Test/LedgerEvents/Gov/InfoAction.hs:135]: Could you make this a function? This is used a number of times in this test, but also in others.

Making it a function would reduce the test's length and make the important part stand out a bit more.

**Comment by carbolymer** [comment]
>One question though: don't we want to use query gov-state (or alike) instead of using foldBlocks at the end? I thought we were moving away from this foldBlock pattern, in its existing shape at least.

@smelc an info action is not visible in the ledger state, not sure about gov-state. I can't use `checkLedgerStateCondition`, because I need to get notified about votes, new proposal, and the proposal getting removed. I can't do that using `checkLedgerStateCondition` function, because it's not returning anything from ledger state, so I can't aggregate on multiple conditions. We should expand `checkLedgerState` with this feature.

File: cardano-testnet/test/cardano-testnet-test/Cardano/Testnet/Test/LedgerEvents/Gov/InfoAction.hs [hunk 4]
```diff
@@ -0,0 +1,413 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE NumericUnderscores #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+
+{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
```
Comment (Jimbo4350) [inline cardano-testnet/test/cardano-testnet-test/Cardano/Testnet/Test/LedgerEvents/Gov/InfoAction.hs:10]: Why do we need this?

File: cardano-testnet/test/cardano-testnet-test/Cardano/Testnet/Test/LedgerEvents/Gov/InfoAction.hs [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+  -- Create DReps -- TODO: Refactor with shelleyKeyGen
+  H.forConcurrently_ [1..3] $ \n -> do
+   H.execCli' execConfig
+     [ "conway", "governance", "drep", "key-gen"
+     , "--verification-key-file", drepVkeyFp n
+     , "--signing-key-file", drepSKeyFp n
+     ]
+
+  -- Create Drep registration certificates
+  let drepCertFile :: Int -> FilePath
```
Comment (Jimbo4350) [inline cardano-testnet/test/cardano-testnet-test/Cardano/Testnet/Test/LedgerEvents/Gov/InfoAction.hs:112]: This is fine for now but we can inject the dreps into the genesis after the next release of api and cli. We will probably want a single standalone test that makes sure the cli can generate the certificates and successfully submit them.

**Review by Jimbo4350 — APPROVED** [review]
LGTM!

File: cardano-testnet/test/cardano-testnet-test/Cardano/Testnet/Test/LedgerEvents/Gov/InfoAction.hs (hunk 4 above)
Comment (carbolymer) [reply cardano-testnet/test/cardano-testnet-test/Cardano/Testnet/Test/LedgerEvents/Gov/InfoAction.hs:10]: afair HLint was complaining about `!! 0`, which is fine here in my opinion

**Pull request merged**